eidas_node/xml.py

```diff
@@ -86,6 +100,55 @@ def decrypt_xml(tree: ElementTree, key_file: str) -> None:
             enc_ctx.decrypt(elm)
 
         remove_extra_xml_whitespace(tree.getroot())
+    return len(encrypted_elements)
+
+
+def encrypt_xml_node(node: Element, cert_file: str, cipher: XmlBlockCipher, key_transport: XmlKeyTransport) -> None:
+    """
+    Encrypt a XML node.
+
+    The node is removed from the parent element and replaced with <EncryptedData> element.
+
+    :param node: A XML subtree.
+    :param cert_file: A path to the certificate file.
+    :param cipher: Encryption algorithm to use.
+    :param key_transport: Key transport algorithm to use.
+    """
+    # Create a template for encryption. xmlsec.template functions don't cover all libxmlsec1 features yet.
+    enc_data = SubElement(node.getparent(),
+                          '{%s}EncryptedData' % XML_ENC_NAMESPACE,
+                          {'Type': xmlsec.EncryptionType.ELEMENT},
+                          nsmap={'xmlenc': XML_ENC_NAMESPACE})
+    SubElement(enc_data, '{%s}EncryptionMethod' % XML_ENC_NAMESPACE, {'Algorithm': cipher.value})
+    SubElement(enc_data, '{%s}CipherData' % XML_ENC_NAMESPACE)
+    xmlsec.template.encrypted_data_ensure_cipher_value(enc_data)
+
+    # Info about the generated encryption key.
+    key_info = xmlsec.template.encrypted_data_ensure_key_info(enc_data, ns='ds')
+    enc_key = SubElement(key_info, '{%s}EncryptedKey' % XML_ENC_NAMESPACE)
+    SubElement(enc_key, '{%s}EncryptionMethod' % XML_ENC_NAMESPACE, {'Algorithm': key_transport.value})
+    SubElement(enc_key, '{%s}CipherData' % XML_ENC_NAMESPACE)
+    xmlsec.template.encrypted_data_ensure_cipher_value(enc_key)
+
+    # Info about the certificate.
+    key_info = xmlsec.template.encrypted_data_ensure_key_info(enc_key, ns='ds')
+    x509_data = xmlsec.template.add_x509_data(key_info)
+    xmlsec.template.x509_data_add_certificate(x509_data)
+    xmlsec.template.x509_data_add_issuer_serial(x509_data)
+
+    # xmlsec library adds unnecessary newlines to the signature template.
+    remove_extra_xml_whitespace(enc_data)
+
+    # Encrypt with a newly generated key
+    key_type, key_length = XML_KEY_INFO[cipher]
```

Comment:
This is not guarding against possible use of unsupported cipher for older xmlsec. I know that this is written in the docs, but probably some `try ... except` should be in order.

Reply:
If the cipher is not supported, it fails at `ctx.encrypt_xml(enc_data, node)` with a generic error without any information about what went wrong. I couldn't find any reliable way how to find out which ciphers are supported or how to get libxmlsec1 version (except for the hack in unit tests).

We could catch `ctx.encrypt_xml(enc_data, node)` exception and reraise it with "Encryption failed. Invalid certificate, invalid or unsupported encryption method.".

Reply:
That sounds reasonable.

Reply:
It now raises `SecurityException`.